src/libexpr/nixexpr-as-json.cc

```diff
@@ -0,0 +1,326 @@
+#include "nixexpr.hh"
+#include "derivations.hh"
+#include "util.hh"
+
+#include <cstdlib>
+
+namespace nix {
+
+// binary operators are implemented in nixexpr.hh MakeBinOp
+
+// https://stackoverflow.com/questions/7724448
+// note: we use full jump table to make this as fast as possible
+// note: we assume valid input. errors should be handled by the nix parser
+// 93 * 7 = 651 byte
+char String_showAsJson_replace_array[93][7] = {
+  "\\u0000", "\\u0001", "\\u0002", "\\u0003", "\\u0004", // 0 - 4
+  "\\u0005", "\\u0006", "\\u0007", "\\b", "\\t", // 5 - 9
+  "\\n", "\\u000b", "\\f", "\\r", "\\u000e", // 10 - 14
+  "\\u000f", "\\u0010", "\\u0011", "\\u0012", "\\u0013", // 15 - 19
+  "\\u0014", "\\u0015", "\\u0016", "\\u0017", "\\u0018", // 20 - 24
+  "\\u0019", "\\u001a", "\\u001b", "\\u001c", "\\u001d", // 25 - 29
+  "\\u001e", "\\u001f", " ", "!", "\\\"", // 30 - 34
+  "#", "$", "%", "&", "'", // 35 - 39
+  "(", ")", "*", "+", ",", // 40 - 44
+  "-", ".", "/", "0", "1", // 45 - 49
+  "2", "3", "4", "5", "6", // 50 - 54
+  "7", "8", "9", ":", ";", // 55 - 59
+  "<", "=", ">", "?", "@", // 60 - 64
+  "A", "B", "C", "D", "E", // 65 - 69
+  "F", "G", "H", "I", "J", // 70 - 74
+  "K", "L", "M", "N", "O", // 75 - 79
+  "P", "Q", "R", "S", "T", // 80 - 84
+  "U", "V", "W", "X", "Y", // 85 - 89
+  "Z", "[", "\\\\", // 90 - 92
+};
+
+void String_showAsJson(std::ostream & o, const std::string & s) {
+  for (auto c = s.cbegin(); c != s.cend(); c++) {
+    if ((std::uint8_t) *c <= 92)
+      o << String_showAsJson_replace_array[(std::uint8_t) *c];
+    else
+      o << *c;
+  }
+}
+
+void Expr::showAsJson(std::ostream & str) const
+{
+    abort();
+}
+
+void ExprInt::showAsJson(std::ostream & str) const
+{
+    str << "{\"type\":\"" << NodeTypeName::ExprInt << "\"";
```

Comment:
Any chance we could use a proper JSON serialization library instead? Nix already uses nlohmann JSON and has a home-grown JSON serialiser in `src/libutil/json.hh`.

Could the interface could be `void ExprType::showAsJson(json& list)` instead and an example implementation could look like this:
```c++
void ExprType::showAsJson(json& list) {
  json elem = {
    { "type", NodeTypeName::ExprType },
    { "value", n },
  };
  list.push_back(elem);
}
```

Reply:
> use a proper JSON serialization library

define proper ... this is working, and its fast
i did not use nlohmann, cos it has a different internal datastructure, iirc

Reply:
> JSON serialiser in `src/libutil/json.hh`

would make an interesting benchmark ... someone, anyone? ; )

Reply:
... i would rather add a machine-readable json format, as i described in https://github.com/NixOS/nix/issues/4726#issuecomment-824993605
benefit: json is smaller, [can be parsed faster](https://github.com/NixOS/nix/pull/4731#issuecomment-831548502)

Reply:
>would make an interesting benchmark ... someone, anyone? ; )
My argument here isn't about performance at all. It is about correctness. It avoid programming errors that we can rule out by not doing the stringy stuff on our own. The nlohmann example also makes the code less verbose IMO.

Reply:
> It is about correctness

can be verified by piping the output to `jq` which throws on json parse errors

```sh
nixpkgs=~/src/nixos/nixpkgs/

parallel --version | grep "GNU parallel" >/dev/null || {
  echo "requires gnu parallel"
  exit 1
}

test_nix_file() {
  f="$1"
  nix="./outputs/out/bin/nix --extra-experimental-features nix-command"
  # fixme: nix prints \r to stderr. workaround: 2>/dev/null
  json="$($nix parse --output-format json "$f" 2>/dev/null)" && {
      echo "$json" | jq >/dev/null && {
          #echo -e "ok $f"
          :
        } || {
          echo -e "json error in $f";
        }
    } || {
      echo -e "nix error in $f"
    }
}
export -f test_nix_file

find $nixpkgs -name '*.nix' | sort --random-sort \
  | parallel --jobs 50 -k --progress test_nix_file
```

nixpkgs has 20K nix files, so this will take some time ...

i found a small bug, fixed in 834f883
apart from that, no json errors so far : )

edgecases are non-ascii strings ... these are simply passed through

```
echo löl | iconv -f utf8 -t latin1
l�l

./outputs/out/bin/nix --extra-experimental-features nix-command parse \
  --expr '"'"$(echo löl | iconv -f utf8 -t latin1)"'"' | jq -r 
l�l
```

Reply:
Again it might work *now* but it requires special care whenever another Expr type is being added or when someone adds a character that you aren't escaping already. It is work that we don't have to do (and maintain). Just use a serializer that is already done.

Reply:
If you care about performance, maybe JSON is not the best format you can aim for. And if you *really* care about performance, you should aim for a lazy event-driven API (the XML parser folks pioneered this and called it "SAX").

Reply:
> Just use a serializer that is already done.

just merge my patch. : P

the whole point of my patch is to provide a FAST nix-to-json parser
and your "pretty" solution via some other high-level json printer is (probably) slower

"probably" = i did no benchmark, but its an educated guess

compare my 30 lines in [src/libexpr/nixexpr-as-json.cc](https://github.com/NixOS/nix/blob/834f883cb51acebce50107f4b9f0d50235e3fc42/src/libexpr/nixexpr-as-json.cc#L11)
(no need to print ascii, no need to validate input)

with 250 lines in nlohmann's `dump_escaped`
https://github.com/nlohmann/json/blob/develop/include/nlohmann/detail/output/serializer.hpp#L238
https://github.com/nlohmann/json/blob/develop/include/nlohmann/detail/output/serializer.hpp#L382

> it requires special care whenever another Expr type is being added

how often does that happen? once in five years?

the only "special care" i can think of is keeping backward compatibility
by appending new types at end-of-file in [src/libexpr/nixexpr-node-types.def](https://github.com/NixOS/nix/blob/834f883cb51acebce50107f4b9f0d50235e3fc42/src/libexpr/nixexpr-node-types.def)

> when someone adds a character that you aren't escaping already

what do you mean?
the json spec is constant
when bugs appear, someone will fix them

Reply:
utf8 test

<details>

```sh
#!/bin/sh

nix="./outputs/out/bin/nix --extra-experimental-features nix-command"

echo 1 test null byte
printf '"a \0 z"' | $nix parse --output-format json /dev/stdin | jq
# error: syntax error, unexpected end of file, expecting '"'

echo 2 test valid utf8
inputString=$'one \1 two \2 three \3 four \4 skull \xE2\x98\xA0 lol \U0001f602'
printf '"%s"' "$inputString" | $nix parse --output-format json /dev/stdin | jq

echo 3 test invalid utf8
# invalid utf8 is ignored by nix parser, so its simply passed through
# https://stackoverflow.com/questions/1301402/example-invalid-utf8-string
inputString=$'\xc3\x28 \xa0\xa1 \xe2\x28\xa1'
printf '"%s"' "$inputString" | $nix parse --output-format json /dev/stdin | jq

echo 4 test random utf8 in 100 byte blocks ... this will loop forever
while true; do
# https://unix.stackexchange.com/questions/245623/how-do-i-create-a-text-file-1-gigabyte-containing-random-characters-with-utf-8
inputString="$(
  dd if=/dev/urandom bs=100 count=1 status=none | perl -CO -ne '
    BEGIN{$/=\4}
    no warnings "utf8";
    print chr(unpack("L>",$_) & 0x7fffffff)
  '
)"
printf '"%s"' "$inputString" | $nix parse --output-format json /dev/stdin | jq >/dev/null || {
  echo "json error in inputString:"
  echo "$inputString" | hexdump -C
}
sleep 0.1 # make the loop easier to kill
done
```

</details>

Reply:
> when someone adds a character that you aren't escaping already

im starting to understand your concern

so, what im currently doing:
escape bytes 0 to 92 = ascii control chars (\n \r \b ...), doublequotes, backslash
everything else (bytes 93 to 255) is simply passed through as is

does this work with unicode?
as for the json spec, it is allowed to have raw unicode in json strings
when unicode is used in json, it must be utf16, for example \u0001

can the escaping break valid unicode?
lets look at the valid unicode byte ranges
ascii bytes are from 0 to 127 = 7 bit
so the first unicode byte must be 128 to 255

what about the following unicode bytes?

https://lemire.me/blog/2018/05/09/how-quickly-can-you-check-that-a-string-is-valid-unicode-utf-8/
converted with [hex2dec.py](https://stackoverflow.com/a/43545026/10440128)

First Byte | Second Byte | Third Byte | Fourth Byte
-- | -- | -- | --
[0,127] |   |   |  
[194,223] | [128,191] |   |  
224 | [160,191] | [128,191] |  
[225,236] | [128,191] | [128,191] |  
237 | [128,159] | [128,191] |  
[238,239] | [128,191] | [128,191] |  
240 | [144,191] | [128,191] | [128,191]
[241,243] | [128,191] | [128,191] | [128,191]
244 | [128,143] | [128,191] | [128,191]

so ... ALL unicode bytes are in the range from 128 to 255
and since i escape only bytes 0 to 92, this works : )

one rare edgecase, where this could break: non-unicode input, for example latin1 encoding.
some ascii control chars, like \1 are encoded as unicode \u0001,
so the result string can be mixed unicode and latin1. simple solution: blame the user.
either for throwing non-unicode strings at nix, or for throwing ascii control chars at nix.
(ascii control chars are worse than non-unicode strings)